002 Sept 25 User Routes

HarshP4585 opened the pull request:
- Reverted the code to follow previous structure.
- Migrated to TypeScript for type safety.

- Affected issue: #17 

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Introduced a new user management router with endpoints for retrieving, creating, updating, and deleting users.
	- Added data transfer objects (DTOs) for user input and output, enhancing data handling.
	- New `User` class implemented to encapsulate user-related data.
	- Enhanced user data structure with additional fields for better tracking.

- **Bug Fixes**
	- Improved error handling for user-related operations.

- **Chores**
	- Updated project configuration to use TypeScript and organized middleware setup.
	- Cleaned up unused files and dependencies to streamline the codebase.
	- Added a newline to the `.gitignore` file for formatting consistency.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

MuhammadKhalilzadeh commented on Servers/package.json:
```diff
@@ -4,19 +4,21 @@
   "scripts": {
     "test": "echo \"Error: no test specified\" && exit 1",
-    "start": "nodemon index.js"
+    "start": "npm run build && node dist/index.js",
```
Thank you for your attention to the details. 🙏🏻
For this part `"npm run build && node dist/index.js"` better to switch back to what you had already, I mean `"nodemon index.js"`.
But, let's not remove this one and maybe create another script like `"dist"` for `"npm run build && node dist/index.js"`

MuhammadKhalilzadeh commented on Servers/src/index.ts:
```diff
@@ -1,16 +1,13 @@
-const app = express();
-const port = 3000;
-
 const cors = require("cors");
-
 const userRouter = require("./routes/users.route");
```
@HarshP4585  Please consider these changes commented by CodeRabbitAI. Since you're using TypeScript, apparently code with ES6 syntax is not ideal.

MuhammadKhalilzadeh commented on Servers/src/index.ts:
```diff
@@ -1,16 +1,13 @@
 const userRouter = require("./routes/users.route");
-const bodyParser = require("body-parser");
+
+const app = express();
+const port = 3000;
```
Thank you for the nice code @HarshP4585 
Now let's make it closer to perfect 👍🏻
And this one is also a very good point!
Are you using `.env` file to have environment variables by the way?

HarshP4585 replied:
> Thank you for the nice code @HarshP4585 Now let's make it closer to perfect 👍🏻 And this one is also a very good point! Are you using `.env` file to have environment variables by the way?

@MuhammadKhalilzadeh , No I'm not using env file, but in the next feature of connecting with postgres, I think I'll require it

MuhammadKhalilzadeh commented on Servers/src/index.ts:
```diff
@@ -1,16 +1,13 @@
 // Middlewares
 app.use(cors());
 app.use(express.json());
-app.use(bodyParser.json())
```
Thanks for applying the changes 🙏🏻

MuhammadKhalilzadeh commented on Servers/src/routes/users.route.ts:
```diff
@@ -0,0 +1,204 @@
+import { UserOut } from "../dtos/userOut.dto";
+import { UserIn } from "../dtos/userIn.dto";
+
+const router = require("express").Router();
+const { writeFile, readFile } = require('fs').promises;
```
Oh! another good point. We should apply this one too!

MuhammadKhalilzadeh commented on Servers/src/routes/users.route.ts:
```diff
@@ -0,0 +1,204 @@
+    const user = users.find(u => {
+
+        for (let p of Object.keys(params)) {
+            if (u[p as keyof User] !== params[p as keyof typeof params]) {
```
@HarshP4585 
What do you think about this suggestion?
Do you think its relevant and can help us?

MuhammadKhalilzadeh commented on Servers/src/routes/users.route.ts:
```diff
@@ -0,0 +1,204 @@
+    try {
+        let users = await getUsers()
+        let body = req.body as UserIn
+        const userIndex = users.map(u => u.id).indexOf(userId)
+        const filteredUser: User = users[userIndex];
```
Error handling is very important!
Please consider this change and also check if there's any other place that needs the same

HarshP4585 replied:
> 

@MuhammadKhalilzadeh , we are already handling this error in existing code in the next lines

MuhammadKhalilzadeh commented on Servers/src/routes/users.route.ts:
```diff
@@ -0,0 +1,204 @@
+
+router.post("/reset-password", async (req: Request, res: Response) => {
+    const { email } = req.query as { email: string }
+    const { password } = req.body as { password: string }
```
Good validation stuff!
Let's apply this one too!

HarshP4585 replied:
> Good validation stuff! Let's apply this one too!

We are throwing 404 Not Found error when we cannot find the user with entered email

MuhammadKhalilzadeh reviewed (changes requested):
Thank you so much for the nice code you've written so far 🙏🏻
You're the best buddy 👍🏻
Would you please apply new changes that I just commented with the help of dear **CodeRabbitAI** 😁
And FYI, the **mock data** for the **User model** is on the way.
So please make sure that the **User model Schema** is the **_same as the link reference I gave you_**.

HarshP4585 commented:
> Thank you so much for the nice code you've written so far 🙏🏻
> You're the best buddy 👍🏻
> Would you please apply new changes that I just commented with the help of dear **CodeRabbitAI** 😁
> And FYI, the **mock data** for the **User model** is on the way.
> So please make sure that the **User model Schema** is the **_same as the link reference I gave you_**.

@MuhammadKhalilzadeh , Thank you very much. I updated the code as per suggested by codersbbitAI and the schema for the user model matches with the mermaid diagram. In recent change, I added the code to hash the password too.

MuhammadKhalilzadeh reviewed (approved):
Thank you for applying the changes 👍🏻
